libraries/eosiolib/contracts/eosio/key_value.hpp

```diff
@@ -0,0 +1,1051 @@
+ *       }
+ * }
+ * @endcode
+ */
+template<typename Class, typename T, eosio::name::raw TableName, eosio::name::raw DbName = eosio::name{"eosio.kvram"}>
+class kv_table {
+
+   enum class kv_it_stat {
+      iterator_ok     = 0,  // Iterator is positioned at a key-value pair
+      iterator_erased = -1, // The key-value pair that the iterator used to be positioned at was erased
+      iterator_end    = -2, // Iterator is out-of-bounds
+   };
+
+public:
+   /**
+    * @ingroup keyvalue
+    *
+    * @brief Defines an index on an EOSIO Key Value Table
+    * @details A Key Value Index allows a user of the table to search based on a given field.
+    * The only restrictions on that field are that it is serializable to a binary representation.
+    * Convenience functions exist to handle most of the primitive types as well as some more complex types, and are
+    * used automatically where possible.
+    */
+   class kv_index {
+      class iterator {
+      public:
+         iterator(eosio::name contract_name, uint32_t itr, kv_it_stat itr_stat, size_t data_size, kv_index* idx) :
+                  contract_name{contract_name}, itr{itr}, itr_stat{itr_stat}, data_size{data_size}, idx{idx} {}
+
+         /**
+          * Returns the value that the iterator points to.
+          * @ingroup keyvalue
+          *
+          * @return The value that the iterator points to.
+          *
+          * Example:
+          * @code
+          * // This assumes the code from the class example.
+          *
+          *    void myaction() {
+          *       // add dan account to table           - see put example
+          *
+          *       auto dan = addresses.index.full_name.find("Dan Larimer");
+          *       eosio::check(dan != addresses.index.full_name.end(), "Dan Larimer is not in the table");
+          *       eosio::check(dan.value().city == "Blacksburg", "Got the wrong value");
+          *    }
+          * @endcode
+          */
+         T value() const {
+            using namespace detail;
+
+            eosio::check(itr_stat != kv_it_stat::iterator_end, "Cannot read end iterator");
+            eosio::check(data_size > 0, "Cannot read a value of size 0");
+
+            uint32_t actual_value_size;
+            uint32_t offset = 0;
+
+            void* buffer = data_size > detail::max_stack_buffer_size ? malloc(data_size) : alloca(data_size);
+            auto stat = internal_use_do_not_use::kv_it_value(itr, offset, (char*)buffer, data_size, actual_value_size);
+
+            eosio::check(static_cast<kv_it_stat>(stat) != kv_it_stat::iterator_end, "Error reading value");
+
+            if (idx->name == idx->tbl->primary_index->name) {
+               datastream<const char*> ds((char*)buffer, actual_value_size);
+
+               T val;
+               ds >> val;
+               if (data_size > detail::max_stack_buffer_size) {
+                  free(buffer);
+               }
+               return val;
```

Comment:
This is duplicated below.  You can reduce the duplication by pulling the extra lookup above this and then making the deserialization unconditional.